src/analysis/history.cpp

```diff
@@ -0,0 +1,69 @@
+// © 2022. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract
+// 89233218CNA000001 for Los Alamos National Laboratory (LANL), which
+// is operated by Triad National Security, LLC for the U.S.
+// Department of Energy/National Nuclear Security Administration. All
+// rights in the program are reserved by Triad National Security, LLC,
+// and the U.S. Department of Energy/National Nuclear Security
+// Administration. The Government is granted for itself and others
+// acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works,
+// distribute copies to the public, perform publicly and display
+// publicly, and to permit others to do so.
+
+#include "history.hpp"
+
+namespace History {
+
+Real ReduceMassAccretionRate(MeshData<Real> *md) {
+  const auto ib = md->GetBoundsI(IndexDomain::interior);
+  const auto jb = md->GetBoundsJ(IndexDomain::interior);
+  const auto kb = md->GetBoundsK(IndexDomain::interior);
+
+  auto pmb = md->GetParentPointer();
+  auto &pars = pmb->packages.Get("geometry")->AllParams();
+  const Real xh = pars.Get<Real>("xh");
+
+  namespace p = fluid_prim;
+  const std::vector<std::string> vars({p::density, p::velocity});
+
+  PackIndexMap imap;
+  auto pack = md->PackVariables(vars, imap);
+
+  const int prho = imap[p::density].first;
+  const int pvel_lo = imap[p::velocity].first;
+  const int pvel_hi = imap[p::velocity].second;
+
+  auto geom = Geometry::GetCoordinateSystem(md);
+
+  Real result = 0.0;
+  parthenon::par_reduce(
+      parthenon::LoopPatternMDRange(), "Phoebus History for Mass Accretion Rate",
+      DevExecSpace(), 0, pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
+        const auto &coords = pack.GetCoords(b);
+        if (coords.x1f(i) <= xh && xh < coords.x1f(i + 1)) {
+          const Real gdet = geom.DetGamma(CellLocation::Cent, k, j, i);
+
+          const Real dx1 = coords.Dx(X1DIR, k, j, i);
+          const Real dx2 = coords.Dx(X2DIR, k, j, i);
+          const Real dx3 = coords.Dx(X3DIR, k, j, i);
+
+          // interp to make sure we're getting the horizon correct
+          auto m = (CalcMassFlux(pack, geom, prho, pvel_lo, pvel_hi, b, k, j, i + 1) -
+                    CalcMassFlux(pack, geom, prho, pvel_lo, pvel_hi, b, k, j, i - 1)) /
+                   (2.0 * dx1);
+          auto flux = (CalcMassFlux(pack, geom, prho, pvel_lo, pvel_hi, b, k, j, i) +
+                       (xh - coords.x1v(i)) * m) *
+                      dx2 * dx3;
```

Comment:
I don't quite follow this equation -- I agree with it when `xh = coords.x1v(i)`, but otherwise I'm confused. Why not just find the two indices `i` and `i+1` that contain `xh` and then weight their `CalcMassFlux` values by `(xh - coords.x1v(i)) / (coords.x1v(i+1) - coords.x1v(i))` and `(coords.x1v(i+1) - xh) / (coords.x1v(i+1) - coords.x1v(i))` respectively?

Reply:
I think that's equivalent... This version is an interpolation of that right?

Reply:
So I used the linear interpolation for `y(xh) = y0 + (xh - x0) m` for slope `m`, which is equivalent to the weighted average form of linear interpolation. A slight difference perhaps, for the slope I used two neighboring cells `i+1` and `i-1`.

Reply:
Yeah youre right sorry, I just wasnt thinking clearly.